File: 011-container-with-most-water.c
// https://leetcode.com/problems/container-with-most-water/

// O(n) time

#define MIN(a,b) (((a)<(b))?(a):(b))
#define MAX(a,b) (((a)>(b))?(a):(b))

int maxArea(int* height, int heightSize) {

	int area = 0;

	int left = 0;
	int right = heightSize - 1;

	while (left < right) {
		area = MAX(area, MIN(height[left], height[right]) * (right - left));

		if (height[left] > height[right]) {
			right--;
		} else {
			left++;
		}
	}

	return area;

}

File: 011-container-with-most-water.js
// https://leetcode.com/problems/container-with-most-water/

// O(n) time

/**
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function (height) {
	var area = 0;

	var left = 0, right = height.length - 1;

	while (left < right) {
		area = Math.max(area, Math.min(height[left], height[right]) * (right - left));

		if (height[left] > height[right]) {
			right--;
		} else {
			left++;
		}
	}

	return area;
};
